File: modules/ai/anthropic.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// ---------------------------------------------------------------------------
// Anthropic / Claude provider
// ---------------------------------------------------------------------------

const defaultAnthropicBaseURL = "https://api.anthropic.com/v1"
const anthropicVersion = "2023-06-01"

// AnthropicProvider implements AiProvider for the Anthropic (Claude) API.
type AnthropicProvider struct {
	config  ProviderConfig
	client  *http.Client
	baseURL string
}

// NewAnthropicProvider creates a new Anthropic provider with the given config.
func NewAnthropicProvider(config ProviderConfig) *AnthropicProvider {
	baseURL := config.BaseURL
	if baseURL == "" {
		baseURL = defaultAnthropicBaseURL
	}
	return &AnthropicProvider{
		config:  config,
		client:  &http.Client{Timeout: config.Timeout()},
		baseURL: baseURL,
	}
}

// Name returns "anthropic".
func (p *AnthropicProvider) Name() string { return "anthropic" }

// ---------------------------------------------------------------------------
// Anthropic wire types
// ---------------------------------------------------------------------------

type anthropicRequest struct {
	Model         string             `json:"model"`
	Messages      []ChatMessage      `json:"messages"`
	System        string             `json:"system,omitempty"`
	MaxTokens     uint32             `json:"max_tokens"`
	Tools         []anthropicTool    `json:"tools,omitempty"`
	Stream        bool               `json:"stream,omitempty"`
	Temperature   *float64           `json:"temperature,omitempty"`
	TopP          *float64           `json:"top_p,omitempty"`
	StopSequences []string           `json:"stop_sequences,omitempty"`
	Thinking      *anthropicThinking `json:"thinking,omitempty"`
}

type anthropicThinking struct {
	Type         string `json:"type"`
	BudgetTokens uint32 `json:"budget_tokens"`
}

type anthropicTool struct {
	Name        string      `json:"name"`
	Description string      `json:"description"`
	InputSchema interface{} `json:"input_schema"`
}

type anthropicResponse struct {
	ID         string             `json:"id"`
	Type       string             `json:"type"`
	Role       string             `json:"role"`
	Content    []anthropicContent `json:"content"`
	Model      string             `json:"model"`
	StopReason string             `json:"stop_reason"`
	Usage      anthropicUsage     `json:"usage"`
}

type anthropicContent struct {
	Type  string          `json:"type"`
	Text  string          `json:"text,omitempty"`
	ID    string          `json:"id,omitempty"`
	Name  string          `json:"name,omitempty"`
	Input json.RawMessage `json:"input,omitempty"`
}

type anthropicUsage struct {
	InputTokens  uint32 `json:"input_tokens"`
	OutputTokens uint32 `json:"output_tokens"`
}

// Streaming wire types
type anthropicStreamEvent struct {
	Type         string                 `json:"type"`
	Index        int                    `json:"index,omitempty"`
	ContentBlock *anthropicContentBlock `json:"content_block,omitempty"`
	Delta        *anthropicDelta        `json:"delta,omitempty"`
	Message      *anthropicMsgDelta     `json:"message,omitempty"`
}

type anthropicContentBlock struct {
	Type  string          `json:"type"`
	ID    string          `json:"id,omitempty"`
	Name  string          `json:"name,omitempty"`
	Input json.RawMessage `json:"input,omitempty"`
	Text  string          `json:"text,omitempty"`
}

type anthropicDelta struct {
	Type        string `json:"type,omitempty"`
	Text        string `json:"text,omitempty"`
	PartialJSON string `json:"partial_json,omitempty"`
	StopReason  string `json:"stop_reason,omitempty"`
}

type anthropicMsgDelta struct {
	StopReason string          `json:"stop_reason,omitempty"`
	Usage      *anthropicUsage `json:"usage,omitempty"`
}

// ---------------------------------------------------------------------------
// ChatCompletion (non-streaming)
// ---------------------------------------------------------------------------

// ChatCompletion sends a non-streaming request to the Anthropic API.
func (p *AnthropicProvider) ChatCompletion(ctx context.Context, messages []ChatMessage, tools []ToolDefinition, opts ChatOptions) (*AiResponse, error) {
	model := opts.Model
	if model == "" {
		model = p.config.DefaultModel
	}

	// Extract system messages from the message list
	systemPrompt, filtered := p.extractSystemPrompt(messages)

	maxTokens := uint32(4096)
	if opts.MaxTokens != nil {
		maxTokens = *opts.MaxTokens
	}
	maxTokens = ensureAnthropicThinkingRoom(maxTokens, opts.ReasoningBudgetTokens)

	reqBody := anthropicRequest{
		Model:         model,
		Messages:      filtered,
		System:        systemPrompt,
		MaxTokens:     maxTokens,
		Temperature:   opts.Temperature,
		TopP:          opts.TopP,
		StopSequences: opts.StopSequences,
	}
	if opts.ReasoningBudgetTokens != nil && *opts.ReasoningBudgetTokens > 0 {
		reqBody.Thinking = &anthropicThinking{Type: "enabled", BudgetTokens: *opts.ReasoningBudgetTokens}
	}

	// Convert tools to Anthropic format (input_schema instead of parameters)
	if len(tools) > 0 {
		reqBody.Tools = p.convertTools(tools)
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to marshal request: %s", err))
	}

	url := p.baseURL + "/messages"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("failed to create request: %s", err))
	}

	p.setHeaders(req)

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("request failed: %s", err))
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, NewAiApiError(
			fmt.Sprintf("API returned status %d: %s", resp.StatusCode, string(respBody)),
			uint16(resp.StatusCode),
		)
	}

	var result anthropicResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to decode response: %s", err))
	}

	return p.convertResponse(&result), nil
}

// ---------------------------------------------------------------------------
// ChatCompletionStream (streaming)
// ---------------------------------------------------------------------------

// ChatCompletionStream sends a streaming request and returns a channel of events.
func (p *AnthropicProvider) ChatCompletionStream(ctx context.Context, messages []ChatMessage, tools []ToolDefinition, opts ChatOptions) (<-chan StreamEvent, error) {
	model := opts.Model
	if model == "" {
		model = p.config.DefaultModel
	}

	systemPrompt, filtered := p.extractSystemPrompt(messages)

	maxTokens := uint32(4096)
	if opts.MaxTokens != nil {
		maxTokens = *opts.MaxTokens
	}
	maxTokens = ensureAnthropicThinkingRoom(maxTokens, opts.ReasoningBudgetTokens)

	reqBody := anthropicRequest{
		Model:         model,
		Messages:      filtered,
		System:        systemPrompt,
		MaxTokens:     maxTokens,
		Stream:        true,
		Temperature:   opts.Temperature,
		TopP:          opts.TopP,
		StopSequences: opts.StopSequences,
	}
	if opts.ReasoningBudgetTokens != nil && *opts.ReasoningBudgetTokens > 0 {
		reqBody.Thinking = &anthropicThinking{Type: "enabled", BudgetTokens: *opts.ReasoningBudgetTokens}
	}

	if len(tools) > 0 {
		reqBody.Tools = p.convertTools(tools)
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, NewAiParseError(fmt.Sprintf("failed to marshal request: %s", err))
	}

	url := p.baseURL + "/messages"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("failed to create request: %s", err))
	}

	p.setHeaders(req)

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, NewAiNetworkError(fmt.Sprintf("request failed: %s", err))
	}

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		return nil, NewAiApiError(
			fmt.Sprintf("API returned status %d: %s", resp.StatusCode, string(respBody)),
			uint16(resp.StatusCode),
		)
	}

	ch := make(chan StreamEvent, 64)
	go p.readStream(resp, ch)
	return ch, nil
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func ensureAnthropicThinkingRoom(maxTokens uint32, thinkingBudget *uint32) uint32 {
	if thinkingBudget == nil || *thinkingBudget == 0 {
		return maxTokens
	}
	if maxTokens > *thinkingBudget {
		return maxTokens
	}
	return *thinkingBudget + 1024
}

func (p *AnthropicProvider) setHeaders(req *http.Request) {
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", p.config.APIKey)
	req.Header.Set("anthropic-version", anthropicVersion)
}

// extractSystemPrompt removes system messages from the list and joins them
// into a single string for the separate "system" field.
func (p *AnthropicProvider) extractSystemPrompt(messages []ChatMessage) (string, []ChatMessage) {
	var systemParts []string
	var filtered []ChatMessage
	for _, msg := range messages {
		if msg.Role == "system" {
			systemParts = append(systemParts, msg.Content)
		} else {
			filtered = append(filtered, msg)
		}
	}
	return strings.Join(systemParts, "\n"), filtered
}

// convertTools converts ToolDefinition slices to Anthropic's tool format.
func (p *AnthropicProvider) convertTools(tools []ToolDefinition) []anthropicTool {
	result := make([]anthropicTool, len(tools))
	for i, t := range tools {
		result[i] = anthropicTool{
			Name:        t.Function.Name,
			Description: t.Function.Description,
			InputSchema: t.Function.Parameters,
		}
	}
	return result
}

func (p *AnthropicProvider) convertResponse(r *anthropicResponse) *AiResponse {
	result := &AiResponse{
		Model: r.Model,
		Usage: TokenUsage{
			PromptTokens:     r.Usage.InputTokens,
			CompletionTokens: r.Usage.OutputTokens,
			TotalTokens:      r.Usage.InputTokens + r.Usage.OutputTokens,
		},
		FinishReason: r.StopReason,
	}

	var contentParts []string
	for _, block := range r.Content {
		switch block.Type {
		case "text":
			contentParts = append(contentParts, block.Text)
		case "tool_use":
			args := "{}"
			if len(block.Input) > 0 {
				args = string(block.Input)
			}
			result.ToolCalls = append(result.ToolCalls, ToolCall{
				ID:   block.ID,
				Type: "function",
				Function: FunctionCall{
					Name:      block.Name,
					Arguments: args,
				},
			})
		}
	}
	result.Content = strings.Join(contentParts, "")

	return result
}

func (p *AnthropicProvider) readStream(resp *http.Response, ch chan<- StreamEvent) {
	defer close(ch)
	defer resp.Body.Close()

	payloads, _ := ParseSSEStream(resp.Body)

	for _, payload := range payloads {
		var evt anthropicStreamEvent
		if err := json.Unmarshal([]byte(payload), &evt); err != nil {
			continue
		}

		switch evt.Type {
		case "content_block_delta":
			if evt.Delta != nil {
				switch evt.Delta.Type {
				case "text_delta":
					ch <- StreamEvent{
						Type:    "content_delta",
						Content: evt.Delta.Text,
					}
				case "input_json_delta":
					// Tool call arguments being streamed
					// We need the content block info from a prior event
					ch <- StreamEvent{
						Type:      "tool_call_delta",
						Arguments: evt.Delta.PartialJSON,
					}
				}
			}

		case "content_block_start":
			if evt.ContentBlock != nil && evt.ContentBlock.Type == "tool_use" {
				ch <- StreamEvent{
					Type:         "tool_call_delta",
					ToolCallID:   evt.ContentBlock.ID,
					ToolCallName: evt.ContentBlock.Name,
				}
			}

		case "message_delta":
			if evt.Delta != nil && evt.Delta.StopReason != "" {
				// Usage info might come with the final delta
				usage := TokenUsage{}
				if evt.Message != nil && evt.Message.Usage != nil {
					usage = TokenUsage{
						PromptTokens:     evt.Message.Usage.InputTokens,
						CompletionTokens: evt.Message.Usage.OutputTokens,
						TotalTokens:      evt.Message.Usage.InputTokens + evt.Message.Usage.OutputTokens,
					}
				}
				ch <- StreamEvent{
					Type:  "usage",
					Usage: &usage,
				}
				ch <- StreamEvent{Type: "done"}
				return
			}

		case "message_stop":
			ch <- StreamEvent{Type: "done"}
			return
		}
	}

	ch <- StreamEvent{Type: "done"}
}

File: packages/ai/src/anthropic.ts
/**
 * @module anthropic
 *
 * Anthropic Claude provider implementation.
 *
 * Adapts the Anthropic Messages API to the AiProvider interface.
 * Handles:
 * - Claude-specific message format (system prompt separate from messages)
 * - Extended thinking / reasoning tokens
 * - Tool use (function calling) in Claude format
 * - Streaming via Server-Sent Events
 * - Rate limiting and error handling
 */
import type { ProviderConfig } from "./provider.js";
import { providerTimeout } from "./provider.js";
import type { ChatMessage, ToolDefinition, ChatOptions, AiResponse, StreamEvent, AiTokenUsage, ToolCall } from "./types.js";
import { newAiParseError, newAiNetworkError, newAiApiError, newAiRateLimitError } from "./error.js";
import { parseSSEStream } from "./stream.js";

/**
 * Anthropic (Claude) Messages API provider.
 *
 * Implements the Anthropic Messages API (POST /v1/messages). Differs from the
 * OpenAI-compatible protocol in three key ways this file adapts for:
 *
 *   1. System prompt is a top-level `system` field, NOT a message in the
 *      `messages` array. extractSystemPrompt() pulls system-role messages out
 *      and joins them into the single `system` string Anthropic expects.
 *
 *   2. Tools are declared with `input_schema` (JSON Schema) rather than the
 *      OpenAI `function.parameters` wrapper. convertTools() reshapes them.
 *
 *   3. Extended thinking: when reasoning_budget_tokens > 0, we send a
 *      `thinking` envelope and must reserve headroom for the thinking budget
 *      inside max_tokens (see ensureAnthropicThinkingRoom).
 *
 * Streaming uses Anthropic's own SSE event types (content_block_delta,
 * content_block_start, message_delta, message_stop), distinct from OpenAI's
 * chunk shape. readStream() maps these onto the shared StreamEvent type.
 */

// ---------------------------------------------------------------------------
// Provider defaults
// ---------------------------------------------------------------------------

/** Default API base when ProviderConfig.baseURL is unset. */
const DEFAULT_ANTHROPIC_BASE_URL = "https://api.anthropic.com/v1";

/** Anthropic API version header (pins the Messages API schema). */
const ANTHROPIC_VERSION = "2023-06-01";

// ---------------------------------------------------------------------------
// Anthropic wire types
// ---------------------------------------------------------------------------

/** Request body for POST /v1/messages. `system` is separate from `messages`. */
interface AnthropicRequest {
  model: string;
  messages: ChatMessage[];
  system?: string;
  max_tokens: number;
  tools?: AnthropicTool[];
  stream?: boolean;
  temperature?: number;
  top_p?: number;
  stop_sequences?: string[];
  thinking?: AnthropicThinking;
}

/** Extended-thinking envelope: type=enabled with a token budget. */
interface AnthropicThinking {
  type: string;
  budget_tokens: number;
}

/** Anthropic tool definition: uses input_schema (JSON Schema), not OpenAI's function wrapper. */
interface AnthropicTool {
  name: string;
  description: string;
  input_schema: unknown;
}

/** Non-streaming Messages response. content[] is a sequence of typed blocks. */
interface AnthropicResponse {
  id: string;
  type: string;
  role: string;
  content: AnthropicContent[];
  model: string;
  stop_reason: string;
  usage: AnthropicUsage;
}

/** One block of a response: "text" or "tool_use" (id/name/input). */
interface AnthropicContent {
  type: string;
  text?: string;
  id?: string;
  name?: string;
  input?: string; // JSON string
}

/** Anthropic usage: input + output tokens (no separate total; we sum them). */
interface AnthropicUsage {
  input_tokens: number;
  output_tokens: number;
}

// Streaming wire types
interface AnthropicStreamEvent {
  type: string;
  index?: number;
  content_block?: AnthropicContentBlock;
  delta?: AnthropicDelta;
  message?: AnthropicMsgDelta;
}

interface AnthropicContentBlock {
  type: string;
  id?: string;
  name?: string;
  input?: string;
  text?: string;
}

interface AnthropicDelta {
  type?: string;
  text?: string;
  partial_json?: string;
  stop_reason?: string;
}

interface AnthropicMsgDelta {
  stop_reason?: string;
  usage?: AnthropicUsage;
}

// ---------------------------------------------------------------------------
// AnthropicProvider class
// ---------------------------------------------------------------------------

/**
 * AnthropicProvider implements the provider interface against the Anthropic
 * Messages API. Handles system-prompt extraction, tool reshaping, extended
 * thinking, and Anthropic-format SSE streaming.
 */
export class AnthropicProvider {
  private config: ProviderConfig;
  private baseURL: string;
  private timeoutMs: number;

  constructor(config: ProviderConfig) {
    this.config = config;
    this.baseURL = config.baseURL || DEFAULT_ANTHROPIC_BASE_URL;
    this.timeoutMs = providerTimeout(config);
  }

  name(): string {
    return "anthropic";
  }

  // -------------------------------------------------------------------------
  // ChatCompletion (non-streaming)
  // -------------------------------------------------------------------------

  /**
   * Sends a non-streaming Messages request and returns the full AiResponse.
   * Extracts the system prompt, reshapes tools, reserves thinking headroom,
   * then parses the content blocks (text + tool_use) into AiResponse.
   */
  async chatCompletion(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    opts: ChatOptions,
  ): Promise<AiResponse> {
    const model = opts.model || this.config.defaultModel;

    // Anthropic requires system content as a top-level field, not a message.
    const { systemPrompt, filtered } = this.extractSystemPrompt(messages);

    // Reserve headroom for the thinking budget if extended thinking is on.
    let maxTokens = opts.max_tokens ?? 4096;
    maxTokens = ensureAnthropicThinkingRoom(maxTokens, opts.reasoning_budget_tokens);

    const reqBody: AnthropicRequest = {
      model,
      messages: filtered,
      system: systemPrompt,
      max_tokens: maxTokens,
      temperature: opts.temperature,
      top_p: opts.top_p,
      stop_sequences: opts.stop_sequences,
    };

    if (opts.reasoning_budget_tokens && opts.reasoning_budget_tokens > 0) {
      reqBody.thinking = { type: "enabled", budget_tokens: opts.reasoning_budget_tokens };
    }

    if (tools.length > 0) {
      reqBody.tools = this.convertTools(tools);
    }

    const body = safeMarshal(reqBody);
    const url = `${this.baseURL}/messages`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(url, {
        method: "POST",
        headers: this.headers(),
        body,
        signal: controller.signal,
      });

      if (!resp.ok) {
        const respBody = await resp.text();
        if (resp.status === 429) {
          const retryAfter = parseInt(resp.headers.get("retry-after") ?? "0", 10);
          throw newAiRateLimitError(`rate limited (429): ${respBody.slice(0, 200)}`, retryAfter);
        }
        throw newAiApiError(
          `API returned status ${resp.status}: ${respBody}`,
          resp.status,
        );
      }

      const result = (await resp.json()) as AnthropicResponse;
      return this.convertResponse(result);
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw newAiNetworkError(`request timed out after ${this.timeoutMs}ms`);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }

  // -------------------------------------------------------------------------
  // ChatCompletionStream (streaming via Anthropic SSE)
  // -------------------------------------------------------------------------

  /**
   * Sends a streaming Messages request. Returns an AsyncIterable<StreamEvent>
   * that yields content_delta, tool_call_delta, usage, and done events as
   * Anthropic SSE events arrive. The HTTP timeout covers the whole stream.
   */
  async chatCompletionStream(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    opts: ChatOptions,
  ): Promise<AsyncIterable<StreamEvent>> {
    const model = opts.model || this.config.defaultModel;
    const { systemPrompt, filtered } = this.extractSystemPrompt(messages);

    let maxTokens = opts.max_tokens ?? 4096;
    maxTokens = ensureAnthropicThinkingRoom(maxTokens, opts.reasoning_budget_tokens);

    const reqBody: AnthropicRequest = {
      model,
      messages: filtered,
      system: systemPrompt,
      max_tokens: maxTokens,
      stream: true,
      temperature: opts.temperature,
      top_p: opts.top_p,
      stop_sequences: opts.stop_sequences,
    };

    if (opts.reasoning_budget_tokens && opts.reasoning_budget_tokens > 0) {
      reqBody.thinking = { type: "enabled", budget_tokens: opts.reasoning_budget_tokens };
    }

    if (tools.length > 0) {
      reqBody.tools = this.convertTools(tools);
    }

    const body = safeMarshal(reqBody);
    const url = `${this.baseURL}/messages`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "POST",
        headers: this.headers(),
        body,
        signal: controller.signal,
      });
    } catch (err) {
      clearTimeout(timer);
      if (err instanceof Error && err.name === "AbortError") {
        throw newAiNetworkError(`request timed out after ${this.timeoutMs}ms`);
      }
      throw err;
    }

    if (!resp.ok) {
      clearTimeout(timer);
      const respBody = await resp.text();
      if (resp.status === 429) {
        const retryAfter = parseInt(resp.headers.get("retry-after") ?? "0", 10);
        throw newAiRateLimitError(`rate limited (429): ${respBody.slice(0, 200)}`, retryAfter);
      }
      throw newAiApiError(
        `API returned status ${resp.status}: ${respBody}`,
        resp.status,
      );
    }

    return this.readStream(resp, timer);
  }

  // -------------------------------------------------------------------------
  // Internal helpers
  // -------------------------------------------------------------------------

  /** Anthropic auth headers: x-api-key + the pinned API version. */
  private headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      "x-api-key": this.config.apiKey,
      "anthropic-version": ANTHROPIC_VERSION,
    };
  }

  /**
   * Removes system-role messages from the list and joins them into one string.
   * Anthropic's API takes `system` as a top-level field rather than a message,
   * so we must split it out of the OpenAI-style messages array before sending.
   * Non-system messages are returned in original order.
   */
  private extractSystemPrompt(
    messages: ChatMessage[],
  ): { systemPrompt: string; filtered: ChatMessage[] } {
    const systemParts: string[] = [];
    const filtered: ChatMessage[] = [];
    for (const msg of messages) {
      if (msg.role === "system") {
        systemParts.push(msg.content ?? "");
      } else {
        filtered.push(msg);
      }
    }
    return { systemPrompt: systemParts.join("\n"), filtered };
  }

  /**
   * Converts provider ToolDefinitions to Anthropic's tool format. Anthropic
   * uses `input_schema` (a raw JSON Schema) instead of OpenAI's
   * `{ function: { parameters } }` wrapper.
   */
  private convertTools(tools: ToolDefinition[]): AnthropicTool[] {
    return tools.map((t) => ({
      name: t.function.name,
      description: t.function.description,
      input_schema: t.function.parameters,
    }));
  }

  /**
   * Parses an Anthropic Messages response into the provider-agnostic AiResponse.
   * Walks content blocks: "text" blocks concatenate into `content`; "tool_use"
   * blocks become ToolCall entries (input is JSON-stringified for the wire).
   * Usage is remapped from input/output_tokens to the shared AiTokenUsage.
   */
  private convertResponse(r: AnthropicResponse): AiResponse {
    const usage: AiTokenUsage = {
      prompt_tokens: r.usage?.input_tokens ?? 0,
      completion_tokens: r.usage?.output_tokens ?? 0,
      total_tokens: (r.usage?.input_tokens ?? 0) + (r.usage?.output_tokens ?? 0),
    };

    const result: AiResponse = {
      content: "",
      tool_calls: [],
      usage,
      model: r.model ?? "",
      finish_reason: r.stop_reason ?? "",
    };

    const contentParts: string[] = [];
    for (const block of r.content ?? []) {
      switch (block.type) {
        case "text":
          contentParts.push(block.text ?? "");
          break;
        case "tool_use": {
          let args = "{}";
          if (block.input && typeof block.input === "string" && block.input.length > 0) {
            args = block.input;
          } else if (block.input && typeof block.input === "object") {
            args = JSON.stringify(block.input);
          }
          result.tool_calls.push({
            id: block.id ?? "",
            type: "function",
            function: {
              name: block.name ?? "",
              arguments: args,
            },
          } satisfies ToolCall);
          break;
        }
      }
    }
    result.content = contentParts.join("");

    return result;
  }

  /**
   * Reads Anthropic's SSE event stream and yields StreamEvents.
   *
   * Event types handled:
   *   - content_block_delta / text_delta:    -> content_delta (streamed text)
   *   - content_block_delta / input_json_delta: -> tool_call_delta (args chunk)
   *   - content_block_start / tool_use:      -> tool_call_delta (id + name)
   *   - message_delta (stop_reason set):     -> usage + done (terminal)
   *   - message_stop:                        -> done (terminal)
   *
   * Unlike OpenAI, tool-call arguments stream as partial_json deltas rather
   * than being accumulated server-side; we forward each fragment directly.
   */
  private async *readStream(resp: Response, timer: ReturnType<typeof setTimeout>): AsyncGenerator<StreamEvent> {
    try {
      if (!resp.body) {
        yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
        return;
      }

      const reader = resp.body.getReader();
      const payloads = await parseSSEStream(reader);

      for (const payload of payloads) {
        let evt: AnthropicStreamEvent;
        try {
          evt = JSON.parse(payload) as AnthropicStreamEvent;
        } catch {
          continue;
        }

        switch (evt.type) {
          case "content_block_delta":
            if (evt.delta) {
              switch (evt.delta.type) {
                case "text_delta":
                  yield {
                    type: "content_delta",
                    content: evt.delta.text ?? "",
                    tool_call_id: "",
                    tool_call_name: "",
                    arguments: "",
                    usage: null,
                  };
                  break;
                case "input_json_delta":
                  // Tool call arguments being streamed
                  yield {
                    type: "tool_call_delta",
                    content: "",
                    tool_call_id: "",
                    tool_call_name: "",
                    arguments: evt.delta.partial_json ?? "",
                    usage: null,
                  };
                  break;
              }
            }
            break;

          case "content_block_start":
            if (evt.content_block?.type === "tool_use") {
              yield {
                type: "tool_call_delta",
                content: "",
                tool_call_id: evt.content_block.id ?? "",
                tool_call_name: evt.content_block.name ?? "",
                arguments: "",
                usage: null,
              };
            }
            break;

          case "message_delta":
            if (evt.delta?.stop_reason) {
              const usage: AiTokenUsage = {
                prompt_tokens: evt.message?.usage?.input_tokens ?? 0,
                completion_tokens: evt.message?.usage?.output_tokens ?? 0,
                total_tokens:
                  (evt.message?.usage?.input_tokens ?? 0) + (evt.message?.usage?.output_tokens ?? 0),
              };
              yield {
                type: "usage",
                content: "",
                tool_call_id: "",
                tool_call_name: "",
                arguments: "",
                usage,
              };
              yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
              return;
            }
            break;

          case "message_stop":
            yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
            return;
        }
      }

      yield { type: "done", content: "", tool_call_id: "", tool_call_name: "", arguments: "", usage: null };
    } finally {
      clearTimeout(timer);
    }
  }
}

/**
 * Ensures max_tokens leaves room for the extended-thinking budget.
 * Anthropic requires max_tokens > thinking.budget_tokens (the output budget
 * must exceed the thinking budget). If the caller's max_tokens is too small,
 * we bump it to budget + 1024. No-op when thinking is disabled (budget=0).
 */
function ensureAnthropicThinkingRoom(maxTokens: number, thinkingBudget?: number): number {
  if (!thinkingBudget || thinkingBudget === 0) {
    return maxTokens;
  }
  if (maxTokens > thinkingBudget) {
    return maxTokens;
  }
  return thinkingBudget + 1024;
}

/** JSON-stringify with a typed error wrapper (guards against circular refs). */
function safeMarshal(obj: unknown): string {
  try {
    return JSON.stringify(obj);
  } catch (err) {
    throw newAiParseError(`failed to marshal request: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/** Creates a new Anthropic provider with the given config. */
export function newAnthropicProvider(config: ProviderConfig): AnthropicProvider {
  return new AnthropicProvider(config);
}
